kernel/packages/decentraland-ecs/src/decentraland/math/Quaternion.ts

```diff
@@ -526,37 +527,54 @@ export class Quaternion {
    * @param up - defines the direction
    */
   public setFromToRotation(from: Vector3, to: Vector3, up: Vector3 = MathTmp.staticUp) {
-    MathTmp.tmpMatrix = Matrix.Zero() // clean up preallocated matrix
-    Matrix.LookAtLHToRef(from, to, up, MathTmp.tmpMatrix)
-    MathTmp.tmpMatrix.invert()
-    Quaternion.FromRotationMatrixToRef(MathTmp.tmpMatrix, this)
+    const result = Quaternion.FromToRotation(from, to)
+    this.x = result.x
+    this.y = result.y
+    this.z = result.z
+    this.w = result.w
   }
 
   /**
-   * Gets the euler angle representation of the rotation.
+   * Sets the euler angle representation of the rotation.
    */
   public set eulerAngles(euler: Vector3) {
     this.setEuler(euler.x, euler.y, euler.z)
   }
 
   /**
-   * Sets the euler angle representation of the rotation.
+   * Gets the euler angle representation of the rotation.
+   * Implemented unity-based calculations from: https://stackoverflow.com/a/56055813
    */
   public get eulerAngles() {
     const out = new Vector3()
-    const mat = new Matrix()
-    this.toRotationMatrix(mat)
-    const m = Matrix.GetAsMatrix3x3(mat)
 
-    out.y = RAD2DEG * Math.asin(Math.max(-1, Math.min(1, m[6])))
+    // if the input quaternion is normalized, this is exactly one. Otherwise, this acts as a correction factor for the quaternion's not-normalizedness
+    const unit = (this.x * this.x) + (this.y * this.y) + (this.z * this.z) + (this.w * this.w)
 
-    if (Math.abs(m[6]) < 0.99999) {
-      out.x = RAD2DEG * Math.atan2(-m[7], m[8])
-      out.z = RAD2DEG * Math.atan2(-m[3], m[0])
-    } else {
-      out.x = RAD2DEG * Math.atan2(m[5], m[4])
+    // this will have a magnitude of 0.5 or greater if and only if this is a singularity case
+    const test = this.x * this.w - this.y * this.z
+
+    if (test > 0.4995 * unit) { // singularity at north pole
+      out.x = Math.PI / 2
+      out.y = 2 * Math.atan2(this.y, this.x)
+      out.z = 0
+    } else if (test < -0.4995 * unit) { // singularity at south pole
+      out.x = -Math.PI / 2
+      out.y = -2 * Math.atan2(this.y, this.x)
       out.z = 0
+    } else { // no singularity - this is the majority of cases
+      out.x = Math.asin(2 * (this.w * this.x - this.y * this.z))
+      out.y = Math.atan2(2 * this.w * this.y + 2 * this.z * this.x, 1 - 2 * (this.x * this.x + this.y * this.y))
+      out.z = Math.atan2(2 * this.w * this.z + 2 * this.x * this.y, 1 - 2 * (this.z * this.z + this.x * this.x))
     }
+    out.x *= RAD2DEG
+    out.y *= RAD2DEG
+    out.z *= RAD2DEG
+
+    // ensure the degree values are between 0 and 360
+    out.x %= 360
+    out.y %= 360
+    out.z %= 360
```

Comment:
This normalization yields strange results, in % is not the modulus operation you may be wanting. It is only the rest of the integer division.

`-97 % 360 = -97` based on your comment you may be expecting it to be `263`

You may be wanting the modulo function:

```ts
function mod(n, p) {
    return n - p * Math.floor( n / p );
}
```

Reply:
however the calculation works great using those % operations but I can test what you propose.

Reply:
If this is just to clamp the angles between 0-360 it should work, BUT modulo doesn't work well for negative numbers, we should be wary about this. Also, using this operator may remove the decimals, and this will incur in loss of precision.

Reply:
Both `mod(3.21, 2)` and `3.21 % 2` yield `1.21` there are no integers in JS. Are you looking to something I am not seeing? @BrianAmadori 

Negative numbers test:
`mod(-11.2, 360) = 348.8` = `360 - 11.2`
`mod(-1001.2, 360) = 78.8` = `-1001.2 + 360+ 360 + 360`

Reply:
Ended up using the `Scalar.Repeat()` method (it does the same caculation as that mod function) because it returns values that are identical to Unity's. (instead of equivalent euler values like `310 and -50`)